### PR TITLE
[python] Let ingest accept `additional_metadata: dict[str, Metadatum]`

### PR DESCRIPTION
**Pull request by ryan-williams**
**Issue and/or context:** #2252 

TODO:
- [x] Add test
- [x] Audit code paths where metadata-writing can be bypassed (e.g.: `skip_existing_nonempty_domain` branches)
- [x] `ingestor.py` tests

### CODE REVIEW

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 1]
```diff
@@ -89,6 +90,9 @@
 _TDBO = TypeVar("_TDBO", bound=TileDBObject[RawHandle])
 
 
+AdditionalMetadata = Optional[Dict[str, str]]
```
[the remark this answers is not in the document]
Comment (johnkerl) [reply apis/python/src/tiledbsoma/io/ingest.py:93]: Yes definitely. I'm sure we'll want `str`, `int`, `float`, `bool`.

File: apis/python/src/tiledbsoma/io/ingest.py [hunk 2]
```diff
@@ -89,6 +90,16 @@
...
+
+
+def add_metadata(
+    obj: TileDBObject[Any], additional_metadata: AdditionalMetadata
+) -> None:
+    for k, v in (additional_metadata or {}).items():
```
Comment (johnkerl) [inline apis/python/src/tiledbsoma/io/ingest.py:99]: check it's open for write and throw an informative error message if it isn't

I believe there's code for that already in `ingest.py` -- either as an already-extracted helper function, or inline code that should be extracted to a helper function

File: apis/python/src/tiledbsoma/io/ingest.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (ryan-williams) [reply apis/python/src/tiledbsoma/io/ingest.py]: done

**Comment by ryan-williams** [comment]
https://github.com/ryan-williams/TileDB-SOMA/pull/1#issuecomment-2012624362: example test including a `tiledb://` URL and new `additional_metadata` passed.

Believe this is g2g.

**Review by johnkerl — APPROVED** [review]
🚢

**Pull request merged**